zkevm-circuits/src/mpt_circuit/account_leaf/account_leaf_nonce_balance.rs

```diff
@@ -58,744 +65,212 @@ The whole account leaf looks like:
-            ));
-
-            /*
-            `c_main.rlp2 = #(nonce bytes) + #(balance bytes) + 32 + 32`.
-            Note that `32 + 32` means the number of codehash bytes + the number of storage root bytes.
-            */
-            constraints.push((
-                "Lean nonce balance c_main.rlp2",
-                q_enable.clone()
-                    * (is_non_existing_account_proof.clone() - is_wrong_leaf.clone() - one.clone())
-                    * (c_rlp2 - nonce_len - balance_len - c66),
-            ));
-
-            /*
-            The whole RLP length of the account leaf is specified in the account leaf key row with
-            `s_main.rlp1 = 248` and `s_main.rlp2`. `s_main.rlp2` in key row actually specifies the length.
-            `s_main.rlp2` in nonce balance row specifies the length of the remaining string in nonce balance
-            row, so we need to check that `s_main.rlp2` corresponds to the key length (in key row) and
-            `s_main.rlp2` in nonce balance row. However, we need to take into account also the bytes
-            where the lengths are stored:
-            `s_main.rlp2 (key row) - key_len - 1 (because key_len is stored in 1 byte)
-                - s_main.rlp2 (nonce balance row) - 1 (because of s_main.rlp1)
-                - 1 (because of s_main.rlp2) = 0`
-
-            Example:
-            [248,106,161,32,252,237,52,8,133,130,180,167,143,97,28,115,102,25,94,62,148,249,8,6,55,244,16,75,187,208,208,127,251,120,61,73,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
-            [184,70,128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,248,68,128,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]
-            We can see: `106 - 33 - 1 - 70 - 1 - 1 = 0`.
-            */
-            constraints.push((
-                "Account leaf RLP length",
-                q_enable
-                    * (is_non_existing_account_proof - is_wrong_leaf - one.clone())
-                    * (rlp_len - key_len - one.clone() - s_rlp2 - one.clone() - one.clone()),
-            ));
-
-            constraints
-        });
-
-        let q_enable_nonce_long = |meta: &mut VirtualCells<F>| {
-            let q_enable = q_enable(meta);
-            let mut is_nonce_long = meta.query_advice(
-                denoter.sel1,
-                Rotation(-(ACCOUNT_LEAF_NONCE_BALANCE_S_IND - ACCOUNT_LEAF_KEY_S_IND)),
-            );
+            require!(account_rlc => rlc);
+
+            // Check the RLP encoding consistency.
+            // RlP encoding: account = [key, [nonce, balance, storage, codehash]]
+            // The only exception is when `is_non_existing_account_proof = 1` &
+            // `is_wrong_leaf = 0`. In this case the value does not matter as
+            // the account leaf is only a placeholder and does not use
+            // `s_main.rlp1` and `s_main.rlp2`.
+            //  TODO(Brecht): Can we remove this if by just making this pass in this special
```

Comment:
The witness generator has been modified to prepare the valid RLP encoding for account leaf placeholder. I prepared a PR with the new tests and simplified code, but I don't have permission to push it.

Reply:
Opened a couple of days ago [here](https://github.com/Brechtpd/zkevm-circuits/pull/6).

Reply:
Oh thanks! Sorry didn't see it before because my notifications were not set correctly.